File: indobase-suite/frappe-app/indobase_suite/indobase_suite/utils/workspace_map.py
"""Deterministic org/project → Workspace team / project keys (mirror bridge/src/workspace-map.ts)."""

from __future__ import annotations

MAX_KEY_LEN = 64


def _clean_slug(value: str) -> str:
	cleaned = "".join(c for c in value.lower() if c.isalnum() or c == "-")
	return cleaned[:MAX_KEY_LEN]


def _clean_project_ref(value: str) -> str:
	cleaned = "".join(c for c in value.lower() if c.isalnum())
	return cleaned[:40]


def workspace_team_key_for_org_slug(org_slug: str) -> str:
	cleaned = _clean_slug(org_slug or "")
	if not cleaned:
		return "ib-ws-org-default"
	return f"ib-ws-org-{cleaned}"[:MAX_KEY_LEN]


def workspace_project_key_for_project_ref(project_ref: str) -> str:
	cleaned = _clean_project_ref(project_ref or "")
	if not cleaned:
		return "ib-ws-proj-default"
	return f"ib-ws-proj-{cleaned}"[:MAX_KEY_LEN]


def build_workspace_map(
	*,
	org_slug: str,
	project_ref: str,
	project_name: str | None = None,
	organization_name: str | None = None,
) -> dict[str, str]:
	team_key = workspace_team_key_for_org_slug(org_slug)
	project_key = workspace_project_key_for_project_ref(project_ref)
	return {
		"org_slug": (org_slug or "").strip(),
		"project_ref": (project_ref or "").strip(),
		"team_key": team_key,
		"project_key": project_key,
		"team_title": (organization_name or org_slug or "Organization")[:140],
		"project_title": (project_name or project_ref or "Project")[:140],
	}


def workspace_home_path(workspace_map: dict[str, str]) -> str:
	return f"/s/{workspace_map['team_key']}/{workspace_map['project_key']}"

File: indobase-suite/bridge/src/workspace-map.ts
/**
 * Deterministic org/project → Workspace team / project keys.
 *
 * Must stay in sync with `frappe-app/indobase_suite/indobase_suite/utils/workspace_map.py`.
 */

export type WorkspaceMap = {
  orgSlug: string
  projectRef: string
  teamKey: string
  projectKey: string
  teamTitle: string
  projectTitle: string
}

const MAX_KEY_LEN = 64

function cleanSlug(input: string): string {
  return input
    .toLowerCase()
    .split('')
    .filter((c) => /[a-z0-9-]/.test(c))
    .join('')
    .slice(0, MAX_KEY_LEN)
}

function cleanProjectRef(input: string): string {
  return input
    .toLowerCase()
    .split('')
    .filter((c) => /[a-z0-9]/.test(c))
    .join('')
    .slice(0, 40)
}

/** Stable team key for an Indobase organization slug. */
export function workspaceTeamKeyForOrgSlug(orgSlug: string): string {
  const cleaned = cleanSlug(orgSlug)
  if (!cleaned) return 'ib-ws-org-default'
  return `ib-ws-org-${cleaned}`.slice(0, MAX_KEY_LEN)
}

/** Stable project key for an Indobase project ref. */
export function workspaceProjectKeyForProjectRef(projectRef: string): string {
  const cleaned = cleanProjectRef(projectRef)
  if (!cleaned) return 'ib-ws-proj-default'
  return `ib-ws-proj-${cleaned}`.slice(0, MAX_KEY_LEN)
}

export function buildWorkspaceMap(opts: {
  orgSlug: string
  projectRef: string
  projectName?: string
  organizationName?: string
}): WorkspaceMap {
  const orgSlug = opts.orgSlug.trim()
  const projectRef = opts.projectRef.trim()
  const teamKey = workspaceTeamKeyForOrgSlug(orgSlug)
  const projectKey = workspaceProjectKeyForProjectRef(projectRef)
  const teamTitle = (opts.organizationName || orgSlug || 'Organization').slice(0, 140)
  const projectTitle = (opts.projectName || projectRef || 'Project').slice(0, 140)

  return {
    orgSlug,
    projectRef,
    teamKey,
    projectKey,
    teamTitle,
    projectTitle,
  }
}

/** Workspace home deep link (bridge path; upstream proxy maps /s/*). */
export function workspaceHomePath(map: WorkspaceMap): string {
  return `/s/${encodeURIComponent(map.teamKey)}/${encodeURIComponent(map.projectKey)}`
}
